Partial ShadowDOM support

ferjm opened the pull request:
This is just an early WIP, not to take it very seriously yet.

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [x] These changes fix #22719 
- [x] There are tests for these changes

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/22743)
<!-- Reviewable:end -->

ferjm committed:
DocumentOrShadowRoot mixin

ferjm committed:
ShadowRoot interface

ferjm committed:
Partial ShadowRoot implementation of DocumentOrShadowRoot

ferjm committed:
Element attachShadow implementation

ferjm committed:
IS_IN_SHADOW_TREE flag

ferjm committed:
Node shadow root owner

ferjm committed:
Expose Element.AttachShadow under dom.shadowdom.enabled pref

emilio commented on components/script/dom/htmlcollection.rs:
```diff
@@ -282,7 +282,14 @@ impl HTMLCollection {
+                if element.upcast::<Node>().is_in_shadow_tree() &&
+                    element.upcast::<Node>().owner_shadow_root() != self.root.owner_shadow_root()
+                {
+                    return false;
```
How can this happen? I don't think it can.

ferjm replied:
Nested shadow trees, but I guess we can ignore that for now.

emilio replied:
I don't think following_nodes digs into shadow trees, right?

ferjm committed:
Test DOM isolation for shadow trees

emilio commented on components/script/dom/macros.rs:
```diff
@@ -632,3 +632,132 @@ macro_rules! handle_potential_webgl_error {
+            let x = *x as f32;
+            let y = *y as f32;
+            let point = &Point2D::new(x, y);
+            let viewport = self.window.window_size().initial_viewport;
```
This needs to retarget nodes in shadow trees and such... Rather than having these implementations duplicated I would have them just in DocumentOrShadowRoot, which also avoids having to implement dummy ShadowRoot::GetBody & co.

emilio commented:
Also, I'm pretty sure you need a flag to know when a node is connected, and replace a bunch of uses of IS_IN_DOCUMENT by that. Also you need to audit all the callers that propagate flags around the tree like the layout dirty flags and change them so they not only look at parent_node but at parent_node_crossing_shadow_root or something

ferjm committed:
Unify DocumentOrShadowRoot implementation

ferjm committed:
Implement shadow-including root, set node as in doc when connected. Makes JS work in shadow trees

ferjm commented:
> Also, I'm pretty sure you need a flag to know when a node is connected, and replace a bunch of uses of IS_IN_DOCUMENT by that

[I ended up setting the IS_IN_DOC flag if the node is connected](https://github.com/servo/servo/pull/22743/commits/d877214c571d72fd5891c1716ec0c721ff15c167#diff-3a91deb90f65260523e1e9bd52154934R292), which I think is essentially the same thing and saves us from replacing some of the uses of IS_IN_DOC by IS_CONNECTED.

emilio commented:
I'd rather not do that, there are some legit uses of IS_IN_DOC / is_in_doc that need to know whether you're on a document or not. Gecko used to do that and we had to rewrite it because there were too many bugs.

ferjm committed:
Add is_connected flag to node and use it to replace most uses of is_in_doc

ferjm committed:
Node retargeting algorithm

ferjm committed:
Make note_dirty_descendants jump around shadow roots

ferjm committed:
Set connected flag not only on elements

ferjm committed:
Bind/unbind shadow host children to/from tree

ferjm committed:
Introduce concept of composed parent node

ferjm committed:
Do not care about shadow roots when getting root element

ferjm committed:
Throw NotSupported when trying to deep clone a shadow root

ferjm committed:
Throw when trying to import or adopt a shadow root

ferjm committed:
Implement concept of shadow including tree order

ferjm committed:
Always get browsing context from document

ferjm committed:
First bits of shadow dom layout

ferjm committed:
Adapt traversals for shadow dom

ferjm committed:
Move stylesheets related code to DocumentOrShadowRoot

ferjm committed:
ShadowRoot stylesheet list

ferjm committed:
Remove stylesheets ownership from DocumentOrShadowRoot

ferjm committed:
Implement TShadowRoot::style_data

ferjm committed:
Expose a way to flush shadow root stylesheets from layout

ferjm committed:
Register/unregister shadow roots in documents when they are connected

ferjm committed:
Invalidate and flush shadow tree stylesheets where needed

ferjm commented:
The shadow tree is now rendered and styled. Shadow tree styles still affect the document though.

<img width="1136" alt="screenshot 2019-02-13 at 16 44 45" src="https://user-images.githubusercontent.com/480202/52725488-9bbbbe00-2fb1-11e9-9df0-4fa413d116bc.png">

ferjm committed:
Do not add shadow tree styles to stylist

ferjm commented:
The last commit fixes the fact that shadow tree styles were applied to the document.

<img width="1136" alt="screenshot 2019-02-13 at 18 24 10" src="https://user-images.githubusercontent.com/480202/52730859-a9c30c00-2fbc-11e9-9615-c09db275ae28.png">

ferjm committed:
Fix Document.Element(s)FromPoint

ferjm committed:
Retarget result of shadowRoot.element(s)FromPoint

ferjm committed:
Return composed parent node. Fixes style sharing panics

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 6809f5830fc66a2c2ba140e487e5f05233646c49 with merge 819191c00c297a09a15505f0ac43d97f90097ad0...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11618)

ferjm committed:
Update test expectations and whitelist ShadowRoot interface

ferjm committed:
Update size of tests with shadow root size

ferjm committed:
Fix formatting issues

ferjm committed:
Shadow DOM layout and style reftest

emilio commented on components/script/dom/element.rs:
```diff
@@ -3232,7 +3336,9 @@ impl Element {
     /// <https://dom.spec.whatwg.org/#connected>
     pub fn is_connected(&self) -> bool {
         let node = self.upcast::<Node>();
```
Why wouldn't this use the new `IS_CONNECTED` flag?

ferjm committed:
Fix the way the IS_CONNECTED flag is set

ferjm committed:
Do not set dirty out-of-doc nodes

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit cc1fdf8cd823e7a3c5ef5ae6233cd475cc7ff965 with merge 9d8d0df8582b413a42325f878528a1f3a68ca1a4...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11623)

ferjm committed:
Remove IS_CONNECTED flag when node is removed from the doc

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 0021c69dd6efa97ddb4c447a02eefe11adfada0d with merge 8fe8f563a7f54a35e0ecd001fe9d2db8f738c694...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11624)

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit e681af0884e0a488903896fc66671ed493e3797d with merge 83014109858504c9715b6df98e98b0151840d46a...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/11681)

ferjm committed:
Update expectations for cssom tests

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 30a1c79407afef7fa3a57bb35eae4e827ba1077f with merge 7f99a9de9e998f71a1c0d28651ebd9676b8c749c...

servo-wpt-sync commented:
Error syncing changes upstream. Logs saved in error-snapshot-1550663780895.

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit c58018a8da246387dd669acf5d0618529995585a with merge 34746cd9072c0f9339d4a47139675dbc4a1d29d0...

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11627), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/11683), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/b-2LzTN3RDW5Kj8D9zr9lQ)
State: approved= try=True

ferjm commented:
We still need to figure out the privileged JS stuff and preferably make styles work when imported from link elements (they currently work from style tags only), but I think this is enough to support the media UI #22720.

r? @emilio

emilio commented on components/layout_thread/dom_wrapper.rs:
```diff
@@ -157,35 +164,78 @@ impl<'ln> NodeInfo for ServoLayoutNode<'ln> {
+impl<'a> PartialEq for ServoShadowRoot<'a> {
+    #[inline]
+    fn eq(&self, other: &ServoShadowRoot) -> bool {
+        self.shadow_root == other.shadow_root
```
Can't you derive this?

emilio commented on components/layout_thread/dom_wrapper.rs:
```diff
@@ -157,35 +164,78 @@ impl<'ln> NodeInfo for ServoLayoutNode<'ln> {
 
-impl<'lr> TShadowRoot for ShadowRoot<'lr> {
-    type ConcreteNode = ServoLayoutNode<'lr>;
+impl<'sr> TShadowRoot for ServoShadowRoot<'sr> {
```
`l` stands for lifetime, so I'd prefer to keep `lr` for consistency with the other `le` and co.

emilio commented on components/layout_thread/dom_wrapper.rs:
```diff
@@ -157,35 +164,78 @@ impl<'ln> NodeInfo for ServoLayoutNode<'ln> {
+        guard: &SharedRwLockReadGuard,
+    ) {
+        unsafe {
+            &self
```
no need for `&`, right?

emilio commented on components/layout_thread/dom_wrapper.rs:
```diff
@@ -411,7 +479,11 @@ impl<'le> TElement for ServoLayoutElement<'le> {
 
     fn traversal_children(&self) -> LayoutIterator<Self::TraversalChildrenIterator> {
-        LayoutIterator(self.as_node().dom_children())
+        LayoutIterator(if let Some(shadow) = self.shadow_root() {
```
This is ok for now. It'd be good to introduce the concept of flat tree in a more formal way, but...

emilio commented on components/layout_thread/dom_wrapper.rs:
```diff
@@ -703,11 +786,20 @@ impl<'le> ::selectors::Element for ServoLayoutElement<'le> {
 
     fn containing_shadow_host(&self) -> Option<Self> {
-        None
+        match self.containing_shadow() {
```
Just `self.containing_shadow().map(|s| s.host())`?

emilio commented on components/layout_thread/lib.rs:
```diff
@@ -1339,6 +1339,17 @@ impl LayoutThread {
+            document.shadow_roots().len()
+        );
+
+        let device = Device::new(MediaType::screen(), initial_viewport, device_pixel_ratio);
```
Minting this is not great, can you use `this.stylist.device()`?

emilio commented on components/layout_thread/lib.rs:
```diff
@@ -1339,6 +1339,17 @@ impl LayoutThread {
+
+        let device = Device::new(MediaType::screen(), initial_viewport, device_pixel_ratio);
+        // Flush shadow roots stylesheets if dirty.
+        for shadow_root in document.shadow_roots() {
```
Doing this in each layout update is unnecessary, can you file a followup to ensure we only do this at least if any shadow root needs the stylesheets flushed?

ferjm replied:
I implementing this in https://github.com/servo/servo/pull/22743/commits/66f84025696c920bbba4648b569ec9fbc4dfb590

emilio commented on components/script/devtools.rs:
```diff
@@ -103,7 +103,7 @@ fn find_node_by_unique_id(
         document
             .upcast::<Node>()
-            .traverse_preorder()
+            .traverse_preorder(/* shadow including */ false)
```
Let's make the argument to `traverse_preorder` an enum instead of commenting the meaning of the boolean. Something like:

```rust
enum ShadowIncluding {
    No,
    Yes,
}
```

emilio commented on components/script/dom/bindings/trace.rs:
```diff
@@ -492,6 +495,8 @@ unsafe_no_jsmanaged_fields!(RenderApiSender);
 unsafe_no_jsmanaged_fields!(ResourceFetchTiming);
 unsafe_no_jsmanaged_fields!(Timespec);
 unsafe_no_jsmanaged_fields!(HTMLMediaElementFetchContext);
+unsafe_no_jsmanaged_fields!(StyleSheetListOwner);
```
I don't think this is sound. `StyleSheetListOwner` can totally have jsmanaged fields, right? Why do you need this?

emilio commented on components/script/dom/cssstyledeclaration.rs:
```diff
@@ -115,10 +115,16 @@ impl CSSStyleOwner {
-                        .Document()
-                        .invalidate_stylesheets();
+                    if let Some(shadow_root) =
+                        shadow_root_from_node(rule.parent_stylesheet().owner().upcast::<Node>())
```
It seems less footgunny to write a function that returns the owner document or shadow root, wdyt?

ferjm replied:
I ended up adding `invalidate_stylesheets` to the `StyleSheetsListOwner` trait and using the `stylesheets_owner_from_node` method instead.

emilio commented on components/script/dom/cssstylerule.rs:
```diff
@@ -118,12 +119,18 @@ impl CSSStyleRuleMethods for CSSStyleRule {
-                .Document()
-                .invalidate_stylesheets();
+            if let Some(shadow_root) =
+                shadow_root_from_node(self.cssrule.parent_stylesheet().owner().upcast::<Node>())
```
You could use that function here as well.

emilio commented on components/script/dom/cssstylesheet.rs:
```diff
@@ -77,10 +83,14 @@ impl CSSStyleSheet {
-                .Document()
-                .invalidate_stylesheets();
+            if let Some(shadow_root) = shadow_root_from_node(self.owner.upcast::<Node>()) {
+                shadow_root.invalidate_stylesheets();
```
And here.

emilio commented on components/script/dom/element.rs:
```diff
@@ -436,6 +443,57 @@ impl Element {
+            &local_name!("p") |
+            &local_name!("section") |
+            &local_name!("span") => {},
+            _ => return Err(Error::NotSupported),
```
You'll need to remove this restriction for internal shadow roots I guess.

emilio reviewed (changes requested):
Very very superficial review for now, need to get to the meat of the changes :)

ferjm committed:
Derive PartialEq for ServoShadowRoot

ferjm committed:
Minor layout thread cleanups:

- lifetime naming
- unrequired reference removal
- containing_shadow_host getter simplification
- use stylist.device getter where possible

ferjm committed:
Flush shadow roots stylesheets only if they changed

ferjm committed:
Introduce ShadowIncluding enum for tree traversals

ferjm committed:
Make StyleSheetListOwner implement JSTraceable

ferjm committed:
Add invalidate_stylesheets to StyleSheetsListOwner trait

ferjm committed:
Relax attachShadow restrictions for user agent widgets

ferjm commented:
r? @emilio
Thanks! I think I addressed all the review comments.

emilio commented on components/layout_thread/dom_wrapper.rs:
```diff
@@ -158,34 +165,69 @@ impl<'ln> NodeInfo for ServoLayoutNode<'ln> {
+        }
+    }
+
+    pub fn flush_stylesheets(
```
I'd prefer for this to be `unsafe`, there's nothing that makes this safe, and this mutates the shadow root's internal state bypassing the refcell.

emilio commented on components/style/dom_apis.rs:
```diff
@@ -275,16 +275,16 @@ where
-    if root.is_in_document() {
-        return root.owner_doc().elements_with_id(id);
-    }
+    if root.is_connected() {
```
This change looks totally unrelated, why? It also deoptimizes the case of calling `getElementById` in a disconnected `ShadowRoot`.

emilio commented on components/style/dom_apis.rs:
```diff
@@ -231,7 +231,7 @@ where
     // is connected to that root.
     if root.as_document().is_some() {
-        debug_assert!(element.as_node().is_in_document(), "Not connected?");
+        debug_assert!(element.as_node().is_connected(), "Not connected?");
```
Loosening this assertion is wrong.

emilio commented on tests/unit/script/size_of.rs:
```diff
@@ -30,10 +30,10 @@ macro_rules! sizeof_checker (
-sizeof_checker!(size_span, HTMLSpanElement, 448);
-sizeof_checker!(size_text, Text, 216);
-sizeof_checker!(size_characterdata, CharacterData, 216);
+sizeof_checker!(size_node, Node, 200);
```
This looks really unfortunate, we should really have a `RareData`-like thing, given Shadow DOM is not too common.

In Gecko all these things are stored in `nsSlots`, in WebKit and Blink in `NodeRareData`.

emilio commented on components/script/dom/documentorshadowroot.rs:
```diff
@@ -0,0 +1,272 @@
+        s: &Arc<Stylesheet>,
+        stylesheets: &mut StylesheetSet<StyleSheetInDocument>,
+    ) {
+        self.window
```
This is not right. `RemoveStylesheet` only cares about document stylesheets.

emilio commented on components/script/dom/documentorshadowroot.rs:
```diff
@@ -0,0 +1,272 @@
+    pub fn add_stylesheet(
+        &self,
+        owner: &Element,
+        stylesheets: &mut StylesheetSet<StyleSheetInDocument>,
```
It's a bit unfortunate to use virtual dispatch here...

Also this doesn't seem to use `self` at all.

emilio commented on components/script/dom/element.rs:
```diff
@@ -2656,7 +2739,7 @@ impl VirtualMethods for Element {
                         None
                     }
                 });
-                if node.is_in_doc() {
```
This is wrong (or rather, incomplete). You don't want to register named elements in the document but in either the document or shadow root. You should be able to add tests for this by adding an element with `name` attribute into a shadowroot.

emilio commented on components/script/dom/node.rs:
```diff
@@ -2625,6 +2777,22 @@ pub fn document_from_node<T: DerivedFrom<Node> + DomObject>(derived: &T) -> DomR
+
+pub fn stylesheets_owner_from_node<T: DerivedFrom<Node> + DomObject>(
+    derived: &T,
+) -> Box<StyleSheetListOwner> {
```
Out of curiosity, why a trait object with virtual dispatch instead of just an enum?

emilio commented on components/style/author_styles.rs:
```diff
@@ -29,6 +30,7 @@ where
     pub data: CascadeData,
     /// The quirks mode of the last stylesheet flush, used because XBL sucks and
     /// we should really fix it, see bug 1406875.
+    #[ignore_malloc_size_of = "XXX"]
```
Just derive this too.

emilio commented on components/script/dom/node.rs:
```diff
@@ -103,6 +109,9 @@ pub struct Node {
     parent_node: MutNullableDom<Node>,
 
+    /// The parent of this node, ignoring shadow roots.
+    composed_parent_node: MutNullableDom<Node>,
```
Why adding the composed parent node here? Seems unfortunate to grow `Node` just for this, when computing using a is straight-forward and would simplify the patch a lot.

emilio commented on components/script/dom/node.rs:
```diff
@@ -118,6 +127,9 @@ pub struct Node {
     /// The document that this node belongs to.
     owner_doc: MutNullableDom<Document>,
 
+    /// The shadow root this node belongs to.
```
Could you document here what does this return for a `ShadowRoot` itself? Does it return `self`? Or something else?

emilio commented on components/script/dom/node.rs:
```diff
@@ -261,14 +280,33 @@ impl Node {
+
+        for node in new_child.traverse_preorder(ShadowIncluding::No) {
+            if parent_in_shadow_tree {
+                if let Some(shadow_root) = self.downcast::<ShadowRoot>() {
```
If owner_shadow_root returns `self` for a `ShadowRoot`, this check is useless.

emilio commented on components/script/dom/node.rs:
```diff
@@ -261,14 +280,33 @@ impl Node {
+                    node.set_owner_shadow_root(&*shadow_root);
+                } else if let Some(shadow_root) = self.owner_shadow_root() {
+                    node.set_owner_shadow_root(&*shadow_root);
+                }
```
We should assert that `owner_shadow_root` is `Some` here, rather than doing nothing if there's broken state.

emilio commented on components/script/dom/htmlbaseelement.rs:
```diff
@@ -63,8 +63,8 @@ impl HTMLBaseElement {
-    pub fn bind_unbind(&self, tree_in_doc: bool) {
-        if !tree_in_doc {
+    pub fn bind_unbind(&self, tree_connected: bool) {
+        if !tree_connected {
```
This is wrong, `<base>` in a shadow tree should do nothing. Can we have a `BindContext` just like we have `UnbindContext` that has both `tree_in_doc` and `tree_connected`?

Again, should be doable to add a test for this.

emilio commented on components/script/dom/htmlbodyelement.rs:
```diff
@@ -149,12 +149,12 @@ impl VirtualMethods for HTMLBodyElement {
+            s.bind_to_tree(tree_connected);
         }
 
-        if !tree_in_doc {
```
This tree_connected check is also wrong, and should keep being `tree_in_doc`.

emilio commented on components/script/dom/htmltitleelement.rs:
```diff
@@ -69,17 +69,17 @@ impl VirtualMethods for HTMLTitleElement {
         }
         let node = self.upcast::<Node>();
-        if tree_in_doc {
+        if tree_connected {
```
Should keep checking for `in_doc`, not connected.

emilio commented on components/script/dom/element.rs:
```diff
@@ -2702,20 +2785,30 @@ impl VirtualMethods for Element {
+
+        if let Some(shadow_root) = self.upcast::<Node>().owner_shadow_root() {
+            let shadow_root = shadow_root.upcast::<Node>();
+            shadow_root.set_flag(NodeFlags::IS_CONNECTED, tree_connected);
```
Touching the shadow root flags here looks wrong, we should do that in `element.rs` when binding the element instead, the same way you remove them.

emilio commented on components/script/dom/element.rs:
```diff
@@ -2702,20 +2785,30 @@ impl VirtualMethods for Element {
 
-        if !tree_in_doc {
+        if !tree_connected {
             return;
```
This is pre-existing, but this looks very wrong? Why would we never call `bind_to_tree` if unconnected? That sounds like a bug.

emilio commented on components/script/dom/element.rs:
```diff
@@ -2730,11 +2823,23 @@ impl VirtualMethods for Element {
 
         let doc = document_from_node(self);
+
+        if self.is_shadow_host() {
```
`if let Some(shadow_root) = ...{`

emilio commented on components/script/dom/htmltitleelement.rs:
```diff
@@ -69,17 +69,17 @@ impl VirtualMethods for HTMLTitleElement {
         let node = self.upcast::<Node>();
-        if node.is_in_doc() {
+        if node.is_connected() {
             node.owner_doc().title_changed();
```
This `is_connected` change is wrong. You don't want a `<title>` in a shadow root to affect the document.

ferjm committed:
Make ServoShadowRoot.flush_stylesheets unsafe

ferjm committed:
Do not send RemoveStyleSheet message for shadow roots

ferjm committed:
Make StylesheetSet an enum instead of a trait object

ferjm committed:
Make StyleSheetListOwner an enum instead of a trait object

ferjm committed:
Derive MallocSizeOf for QuirksMode

ferjm committed:
Do not store composed parent node

ferjm committed:
Document owner_shadow_root

ferjm committed:
Register named elements in either the document or shadow tree

ferjm committed:
Introduce BindContext with in_doc and connected flags

Fix some is_in_doc -> is_connected mistakes

ferjm committed:
Introduce NodeRareData and ElementRareData

ferjm committed:
Move mutation observers list to NodeRareData

emilio commented on components/script/dom/document.rs:
```diff
@@ -418,6 +375,10 @@ pub struct Document {
     /// https://html.spec.whatwg.org/multipage/#completely-loaded
     completely_loaded: Cell<bool>,
+    /// List of shadow roots bound to the document tree.
+    shadow_roots: DomRefCell<Vec<Dom<ShadowRoot>>>,
```
Given this doesn't need any order, it may as well be a HashSet, but maybe followup if it's hard.

emilio commented on components/script/dom/element.rs:
```diff
@@ -2703,22 +2799,35 @@ impl VirtualMethods for Element {
         }
 
-        if !tree_in_doc {
+        if let Some(shadow_root) = self.upcast::<Node>().owner_shadow_root() {
```
This should call document.register_shadow_root, too.

emilio commented on components/script/dom/element.rs:
```diff
@@ -436,6 +450,59 @@ impl Element {
+            .shadow_root
+            .or_init(|| ShadowRoot::new(self, &*self.node.owner_doc()));
+
+        self.node.owner_doc().register_shadow_root(&*shadow_root);
```
This should only register it if the host is connected (and should probably also set the connected bit on the ShadowRoot, though that's done in Shadowroot::new so fine).

emilio commented on components/script/dom/document.rs:
```diff
@@ -3283,6 +3141,95 @@ impl Document {
+
+    pub fn unregister_shadow_root(&self, shadow_root: &ShadowRoot) {
+        let mut shadow_roots = self.shadow_roots.borrow_mut();
+        let position = shadow_roots.iter().position(|sr| **sr == *shadow_root);
```
This should probably at least `debug_assert!(position.is_some())`.

emilio commented on components/layout_thread/dom_wrapper.rs:
```diff
@@ -375,6 +423,33 @@ impl<'ld> ServoLayoutDocument<'ld> {
+            self.document
+                .shadow_roots()
+                .iter()
+                .map(|sr| ServoShadowRoot::from_layout_js(*sr))
```
This should probably debug_assert!(sr.is_connected()).

ferjm committed:
Add custom elements related stuff into ElementRareData

ferjm committed:
Update size of tests after *RareData changes

ferjm committed:
Set IS_CONNECTED flag on host and children instead of on containing shadow root

ferjm committed:
Tweak list of shadow roots attached to doc

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 84e10d7b651c860aeefa95aa98833745eeafd3c9 with merge d85636917ec6f5034aaebcfd6a9736fa7c543d1d...

ferjm commented:
I think I addressed all review feedback. r? @emilio. Thanks!

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/11771)

jdm commented:
That's #13480 again.

emilio commented on components/layout_thread/dom_wrapper.rs:
```diff
@@ -985,6 +1084,13 @@ impl<'ln> ThreadSafeLayoutNode for ServoThreadSafeLayoutNode<'ln> {
     }
 
     fn children(&self) -> LayoutIterator<Self::ChildrenIterator> {
+        if let Some(element) = self.node.as_element() {
```
nit: This may be more concise with and_then

emilio commented on components/script/devtools.rs:
```diff
@@ -103,7 +103,7 @@ fn find_node_by_unique_id(
         document
             .upcast::<Node>()
-            .traverse_preorder()
+            .traverse_preorder(ShadowIncluding::No)
```
This probably wants to be shadow_including, if it's for devtools, but can be a followup, with its own tests and such.

emilio commented on components/script/dom/document.rs:
```diff
@@ -418,6 +375,10 @@ pub struct Document {
     delayed_tasks: DomRefCell<Vec<Box<dyn TaskBox>>>,
     /// https://html.spec.whatwg.org/multipage/#completely-loaded
     completely_loaded: Cell<bool>,
+    /// List of shadow roots bound to the document tree.
```
No longer a list, so s/List/Set? :) 

Also maybe s/bound/connected?

emilio commented on components/script/dom/documentorshadowroot.rs:
```diff
@@ -0,0 +1,323 @@
+        // Limit the scope of the borrow because id_map might be borrowed again by
+        // GetElementById through the following sequence of calls
+        // reset_form_owner_for_listeners -> reset_form_owner -> GetElementById
+        {
```
I don't think this is relevant now that this is on its own function, so I'd remove the comment and the scope.

emilio commented on components/script/dom/documentorshadowroot.rs:
```diff
@@ -0,0 +1,323 @@
+        // Limit the scope of the borrow because id_map might be borrowed again by
+        // GetElementById through the following sequence of calls
+        // reset_form_owner_for_listeners -> reset_form_owner -> GetElementById
+        {
```
ditto

emilio commented on components/script/dom/documentorshadowroot.rs:
```diff
@@ -0,0 +1,323 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
I'm assuming most of this file is moving code, so I'm not reviewing that code in much detail. Hopefully that's fine.

emilio commented on components/style/dom_apis.rs:
```diff
@@ -276,15 +277,15 @@ where
     if root.is_in_document() {
-        return root.owner_doc().elements_with_id(id);
-    }
+        if let Some(shadow) = root.as_shadow_root() {
```
This is still wrong and doesn't seem addressed. is_in_document() and as_shadow_root().is_some() are mutually exclusive. This should be reverted.

emilio commented on components/style/dom_apis.rs:
```diff
@@ -241,10 +241,11 @@ where
     }
 
     if root.as_shadow_root().is_some() {
+        debug_assert!(element.as_node().is_in_document(), "Not connected?");
```
This change I don't understand either. Please revert or explain how it's supposed to hold.

emilio commented on components/style/dom.rs:
```diff
@@ -188,6 +188,9 @@ pub trait TNode: Sized + Copy + Clone + Debug + NodeInfo + PartialEq {
     fn is_in_document(&self) -> bool;
 
+    /// Returns whether the node is connected.
+    fn is_connected(&self) -> bool;
```
This doesn't seem to be used in the style crate, can it be moved to layout?

emilio commented on components/style/stylesheet_set.rs:
```diff
@@ -375,69 +375,137 @@ where
-            if let Some(device) = device {
-                self.invalidations.collect_invalidations_for(device, sheet, guard);
+macro_rules! stylesheetset_impl {
+    ($set_name:expr, $set_type:ty) => {
```
Given this is no longer a trait, why does this need to change?

emilio commented on tests/unit/script/size_of.rs:
```diff
@@ -30,10 +30,10 @@ macro_rules! sizeof_checker (
-sizeof_checker!(size_span, HTMLSpanElement, 448);
-sizeof_checker!(size_text, Text, 216);
-sizeof_checker!(size_characterdata, CharacterData, 216);
+sizeof_checker!(size_node, Node, 160);
```
Ohai, this is much nicer ;)

emilio commented on components/script/dom/range.rs:
```diff
@@ -759,7 +760,11 @@ impl RangeMethods for Range {
         // Step 11
         let new_offset = new_offset +
-            if node.type_id() == NodeTypeId::DocumentFragment {
+            if node.type_id() ==
```
Can this just be `if let NodeTypeId::DocumentFragment(_) = node.type_id() {`?

emilio commented on components/script/dom/node.rs:
```diff
@@ -303,17 +329,18 @@ impl Node {
             );
         }
-        for node in child.traverse_preorder() {
+        for node in child.traverse_preorder(ShadowIncluding::Yes) {
```
(we can probably merge the two loops, maybe followup)

ferjm replied:
There's [a comment](https://github.com/servo/servo/pull/22743/files/84e10d7b651c860aeefa95aa98833745eeafd3c9#diff-3a91deb90f65260523e1e9bd52154934R344) explaining why the two loops are required.

emilio commented on components/script/dom/node.rs:
```diff
@@ -852,11 +891,25 @@ impl Node {
-    pub fn inclusive_ancestors(&self) -> impl Iterator<Item = DomRoot<Node>> {
-        SimpleNodeIterator {
+    pub fn inclusive_ancestors(&self) -> Box<Iterator<Item = DomRoot<Node>>> {
+        Box::new(SimpleNodeIterator {
```
Why is the box needed? It's a bit unfortunate.

emilio commented on components/script/dom/node.rs:
```diff
@@ -1033,6 +1107,27 @@ impl Node {
+        let mut a = DomRoot::from_ref(&*self);
+        loop {
+            // Step 1.
+            let a_root = a.GetRootNode(&GetRootNodeOptions::empty());
```
You should make sure to optimize `GetRootNode` in a shadow tree to be fast and return the containing shadow, and similar for nodes in the document to return the document.

emilio commented on components/script/dom/node.rs:
```diff
@@ -1118,7 +1214,13 @@ impl LayoutNodeHelpers for LayoutDom<Node> {
     #[inline]
     #[allow(unsafe_code)]
     unsafe fn parent_node_ref(&self) -> Option<LayoutDom<Node>> {
```
This should probably be renamed to `composed_parent_node_ref`.

emilio commented on components/script/dom/node.rs:
```diff
@@ -1416,7 +1536,13 @@ impl TreeIterator {
     fn next_skipping_children_impl(&mut self, current: DomRoot<Node>) -> Option<DomRoot<Node>> {
-        for ancestor in current.inclusive_ancestors() {
+        let iter = if self.shadow_including {
+            current.shadow_including_inclusive_ancestors()
```
Ah, I guess this is why you need the Box<>. Would be nice to either box here, or not box at all and make both iterators have the same type. Again can be followup.

emilio commented on components/script/dom/node.rs:
```diff
@@ -2110,13 +2250,19 @@ impl NodeMethods for Node {
+                return shadow_root.Host().upcast::<Node>().GetRootNode(options);
+            }
+        }
         self.inclusive_ancestors().last().unwrap()
```
If we're in the document this should return the document directly as an optimization.

emilio reviewed (changes requested):
Looks pretty close! Just some questions about the `style/` changes and some other nits really.

ferjm committed:
Minor tweaks: rename composed_parent_node_ref, remove or update outdated comments...

ferjm committed:
Revert style/dom_apis changes

ferjm committed:
Move is_connected function from style to layout, where it is used

ferjm committed:
Make devtools find node by unique id function include shadow trees

ferjm committed:
Revert changes in sheet_set_methods macro

ferjm committed:
Merge Node::shadow_including_inclusive_ancestors into inclusive_ancestors

ferjm committed:
Optimize Node::GetRootNode

ferjm committed:
Update tests manifest

ferjm commented:
Done :) r? @emilio

emilio commented on components/style/stylesheet_set.rs:
```diff
@@ -383,61 +445,62 @@ where
-        ) {
-            if let Some(device) = device {
-                self.invalidations.collect_invalidations_for(device, sheet, guard);
+            fn collect_invalidations_for(
```
This looks misindented now?

emilio commented on components/style/gecko/wrapper.rs:
```diff
@@ -411,6 +411,11 @@ impl<'ln> TNode for GeckoNode<'ln> {
     }
 
+    #[inline]
+    fn is_connected(&self) -> bool {
```
nit: I think this should go away now.

emilio commented on components/script/dom/node.rs:
```diff
@@ -118,6 +124,9 @@ pub struct Node {
     owner_doc: MutNullableDom<Document>,
 
+    /// Rare node data.
+    rare_data: Box<NodeRareData>,
```
Oh, I totally missed this, the point of this is that this should be an `Option<Box<>>` (or such), and be `None` when not needed. otherwise this doesn't help reduce memory usage at all, mostly the opposite.

emilio reviewed (commented):
Sorry for overlooking the `RareData` stuff earlier, it really needs to be lazily allocated, otherwise there's no win at all.

I think with that fixed apart from the other minor nits, this should be good to go.

ferjm committed:
Make Node and Element rare_data an Option

ferjm commented:
I think this is ready now. r? @emilio

bors-servo commented:
:umbrella: The latest upstream changes (presumably #23013) made this pull request unmergeable. Please resolve the merge conflicts.

emilio commented on components/script/dom/node.rs:
```diff
@@ -546,26 +597,24 @@ impl Node {
-                .unwrap()
-                .restyle(damage),
+            NodeTypeId::CharacterData(CharacterDataTypeId::Text) => {
+                if let Some(parent) = self.composed_parent_node() {
```
This should use instead:

`self.parent_node.get().unwrap().dirty(damage)`,

And this match should handle `ShadowRoot`, deferring to the host. That should prevent the panics you were seeing.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #22891) made this pull request unmergeable. Please resolve the merge conflicts.

ferjm committed:
Set dirty descendants if node is connected

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 425ce7addd2d19aabfb89961a2b1d9ad03a4ab24 with merge cb862d967ab26f3e94195a0205afa45abce4cf64...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11818)

ferjm committed:
Add dom.shadowdom.enabled to preferences list

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 769b96795b2c34b2023e4ea83df05b35c4eeaefd with merge 1e4666375743c00621e22cfb91e75164f05725e4...

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11824), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/11880), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/ESM1UwbVTLOXuNFKtF0VPA)
State: approved= try=True

ferjm committed:
Set dirty descendants flag only for elements

ferjm commented:
r? @emilio

emilio commented on components/script/dom/macros.rs:
```diff
@@ -632,3 +632,33 @@ macro_rules! handle_potential_webgl_error {
+        }
+
+        fn rare_data(&self) -> Ref<Option<Box<$type>>> {
+            self.init_rare_data();
```
This doesn't make sense. The point of the rare data is that unless you need to initialize something in there you should not allocate it. So no need to call `init_rare_data` here. Callers of course need to deal with it...

emilio commented on components/script/dom/macros.rs:
```diff
@@ -632,3 +632,33 @@ macro_rules! handle_potential_webgl_error {
+
+macro_rules! impl_rare_data (
+    ($type:ty) => (
+        fn init_rare_data(&self) {
```
This should just be inlined in `rare_data_mut`, and `rare_data_mut` should probably be called `ensure_rare_data` (to highlight the lazy initialization) and should not return an Option.

emilio commented on components/script/dom/macros.rs:
```diff
@@ -632,3 +632,33 @@ macro_rules! handle_potential_webgl_error {
+
+        fn rare_data_mut(&self) -> RefMut<Option<Box<$type>>> {
+            self.init_rare_data();
+            self.rare_data.borrow_mut()
```
This should return `RefMut<Box<$type>>`.

emilio reviewed (changes requested):
I have a few concerns. One is with the `owner_shadow_root` terminology. I don't think that terminology reflects what the functions return, which is just the containing shadow root.

The other one is with the rare data implementation, which right now doesn't accomplish what it's supposed to accomplish, since it's initialized lazily as soon as you try to read from it.

ferjm committed:
Revert unnecessary format changes

ferjm committed:
Rename StylesheetSet to StylesheetSetRef

ferjm committed:
Rename shadow_root_from_node to containing_shadow_root

ferjm committed:
Move StylesheetSetRef to script

ferjm committed:
Do not lazy initialize RareData on its getters

ferjm commented:
r? @emilio

emilio commented on components/script/dom/element.rs:
```diff
@@ -2826,13 +2811,15 @@ impl VirtualMethods for Element {
-            for node in shadow_root.children() {
-                node.set_flag(NodeFlags::IS_CONNECTED, context.tree_connected);
-                node.bind_to_tree(context);
+        if let Some(rare_data) = self.rare_data().as_ref() {
```
a `shadow_root` getter would be neat.

emilio commented on components/script/dom/node.rs:
```diff
@@ -2293,14 +2293,21 @@ impl NodeMethods for Node {
-            DomRoot::from_ref(shadow_root.upcast::<Node>())
-        } else if self.is_in_doc() {
+
+        if let Some(rare_data) = self.rare_data().as_ref() {
```
`if let Some(shadow) = self.containing_shadow_root()`?

emilio commented on components/script/dom/node.rs:
```diff
@@ -2293,14 +2293,21 @@ impl NodeMethods for Node {
-                // shadow-including root.
-                return shadow_root.Host().upcast::<Node>().GetRootNode(options);
+            if let Some(rare_data) = self.rare_data().as_ref() {
+                if let Some(ref shadow_root) = rare_data.owner_shadow_root {
```
`if let Some(shadow) = self.containing_shadow_root()`?

emilio commented on components/script/dom/node.rs:
```diff
@@ -1281,13 +1281,13 @@ impl LayoutNodeHelpers for LayoutDom<Node> {
     #[inline]
     #[allow(unsafe_code)]
     unsafe fn owner_shadow_root_for_layout(&self) -> Option<LayoutDom<ShadowRoot>> {
```
`containing_shadow_root`?

emilio commented on components/script/dom/node.rs:
```diff
@@ -1281,13 +1281,13 @@ impl LayoutNodeHelpers for LayoutDom<Node> {
-            .owner_shadow_root
-            .as_ref()
-            .map(|sr| sr.to_layout())
+        if let Some(rare_data) = (*self.unsafe_get()).rare_data_for_layout().as_ref() {
```
You can also use `?` to make this a bit more compact if you want.

emilio commented on components/script/dom/node.rs:
```diff
@@ -951,16 +950,17 @@ impl Node {
         if let Some(ref shadow_root) = self.downcast::<ShadowRoot>() {
```
`owner_shadow_root` -> `containing_shadow_root`? And why is this special-case needed? The containing shadow root's rare data pointer should point to itself, then you don't need it.

Otherwise you also need it on the layout accessors.

emilio commented on components/script/dom/element.rs:
```diff
@@ -458,7 +443,10 @@ impl Element {
 
     pub fn is_shadow_host(&self) -> bool {
-        self.rare_data().as_ref().unwrap().shadow_root.is_some()
+        if let Some(rare_data) = self.rare_data().as_ref() {
```
With a `shadow_root` accessor this can be `self.shadow_root().is_some()` which reads much nicer imo.

emilio reviewed (changes requested):
Looks much closer!

`owner_shadow_root` still wasn't renamed to `containing_shadow_root`, and I'm 99.99% sure the custom elements code is unsound now.

I have a few other nits in comments, but nothing major I think.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #23263) made this pull request unmergeable. Please resolve the merge conflicts.

ferjm committed:
Final nits; fix custom elements rare data usage; s/owner_s_r/containing_s_r

Clarify special case for containing_shadow_root and add it to layout
accessor

ferjm commented:
r? @emilio

emilio commented on components/script/dom/element.rs:
```diff
@@ -332,75 +332,60 @@ impl Element {
-            .unwrap()
-            .custom_element_definition
-            .clone()
+        if let Some(rare_data) = self.rare_data().as_ref() {
```
You could use: `self.rare_data().as_ref()?.custom_element_definition.clone();`

emilio commented on components/script/dom/element.rs:
```diff
@@ -337,49 +332,63 @@ impl Element {
 
     pub fn clear_reaction_queue(&self) {
-        self.custom_element_reaction_queue.borrow_mut().clear();
+        self.ensure_rare_data()
```
This shouldn't call `ensure_rare_data`. If there's no rare data there's nothing to clear.

So `if let Some(data) = self.rare_data_mut() { data.custom_element_reaction_queue.clear() }` or such.

emilio commented on components/script/dom/element.rs:
```diff
@@ -436,6 +445,69 @@ impl Element {
     }
+
+    fn shadow_root(&self) -> Option<DomRoot<ShadowRoot>> {
+        return self
```
no need for return.

emilio commented on components/script/dom/node.rs:
```diff
@@ -422,14 +452,26 @@ impl Node {
+    /// initialize the raredata.
     pub fn registered_mutation_observers(&self) -> RefMut<Vec<RegisteredObserver>> {
-        self.mutation_observers.borrow_mut()
+        RefMut::map(self.ensure_rare_data(), |rare_data| {
```
Ouch, yeah, I missed the details of this always allocating the rare data even if empty. Please file an issue to fix this.

emilio commented on components/script/dom/node.rs:
```diff
@@ -876,12 +946,29 @@ impl Node {
     }
 
+    pub fn containing_shadow_root(&self) -> Option<DomRoot<ShadowRoot>> {
+        // NodeRareData contains the shadow root the node belongs to,
```
Mind explaining how can a shadow root belong to another shadow root? Can we instead make a `shadow_root.set_containing_shadow_root(shadow_root)` in `AttachShadow`, and then remove this and the similar special-cases? I think that'd be nicer.

emilio commented on components/script/dom/node.rs:
```diff
@@ -2129,14 +2287,31 @@ impl NodeMethods for Node {
+        }
+
+        if let Some(shadow_root) = self.containing_shadow_root() {
+            return DomRoot::from_ref(shadow_root.upcast::<Node>());
```
May be worth simplifying this moving the `if options.composed` here, and removing the block above, so:

```rust
if let Some(shadow_root) = self.containing_shadow_root() {
    return if options.composed {
        shadow_root.Host().upcast::<Node>().GetRootNode(options);
    } else {
        DomRoot::from_ref(shadow_root.upcast())
    }
}
```

emilio reviewed (approved):
r=me with followups filed for:

 * Sharing storage for rare datas.
 * Fix the mutation observer stuff so that we don't always allocate rare data.

And the stuff I point out below fixed. Thanks for bearing with me, and sorry it took quite long :)

ferjm committed:
Set self as containing_shadow_root for shadow roots

ferjm commented:
> Thanks for bearing with me, and sorry it took quite long :)

Thank **you** for your mentoring and patient :)

ferjm commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit 37e88e7 has been approved by `emilio`

<!-- @bors-servo r=emilio 37e88e77cdf00e3555599dd4004d03548bd95dcf -->

bors-servo commented:
:hourglass: Testing commit 37e88e77cdf00e3555599dd4004d03548bd95dcf with merge 30ded91a51073032efe222440e1bddd52e8fb8df...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12077)

ferjm commented:
@bors-servo retry

bors-servo commented:
:bomb: Failed to start rebuilding: `Unknown error`

bors-servo commented:
:hourglass: Testing commit 37e88e77cdf00e3555599dd4004d03548bd95dcf with merge 799490a02e9bea575bf34c39f045ef0883539f05...

bors-servo commented:
:sunny: Test successful - [android-mac](http://build.servo.org/builders/android-mac/builds/1483), [arm32](http://build.servo.org/builders/arm32/builds/14362), [arm64](http://build.servo.org/builders/arm64/builds/14390), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12080), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/12139), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/6978), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/6936), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/11173), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/11240), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/6992), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/6916), [magicleap](http://build.servo.org/builders/magicleap/builds/1101), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/X4EYO1XHSA-B4LA0NytX2w)
Approved by: emilio
Pushing 799490a02e9bea575bf34c39f045ef0883539f05 to master...
